tree/ntuple/v7/inc/ROOT/RClusterPool.hxx

```diff
@@ -0,0 +1,128 @@
+#include <ROOT/RPageStorage.hxx> // for ColumnSet_t
+
+#include <condition_variable>
+#include <memory>
+#include <mutex>
+#include <future>
+#include <queue>
+#include <thread>
+#include <set>
+#include <vector>
+
+namespace ROOT {
+namespace Experimental {
+namespace Detail {
+
+class RPageSource;
+
+// clang-format off
+/**
+\class ROOT::Experimental::Detail::RClusterPool
+\ingroup NTuple
+\brief Managed a set of clusters containing compressed and packed pages
+
+The cluster pool steers the preloading of (partial) clusters.
+*/
+// clang-format on
+class RClusterPool {
+private:
+   /// Maximum number of queued cluster requests for the I/O thread. A single request can span mutliple clusters.
+   static constexpr unsigned int kWorkQueueLimit = 4;
+
+   /// Request to load a subset of the columns of a particular cluster.
+   /// Work items come in groups and are executed by the page source.
+   struct RWorkItem {
+      std::promise<std::unique_ptr<RCluster>> fPromise;
+      DescriptorId_t fClusterId = kInvalidDescriptorId;
+      RPageSource::ColumnSet_t fColumns;
+      RWorkItem() = default;
+   };
+
+   /// Clusters that are currently being processed by the I/O thread.  Every in flight cluster has a corresponding
+   /// work item.
+   struct RInFlightCluster {
+      std::future<std::unique_ptr<RCluster>> fFuture;
+      DescriptorId_t fClusterId = kInvalidDescriptorId;
+      RPageSource::ColumnSet_t fColumns;
+      /// By the time a cluster has been loaded, it might not be necessary anymore.
+      bool fIsExpired = false;
+      RInFlightCluster() = default;
+      bool operator== (const RInFlightCluster &other) const { return fClusterId == other.fClusterId && fColumns == other.fColumns; }
+      bool operator!= (const RInFlightCluster &other) const { return !(*this == other); }
+      /// First order by cluster id, then by number of columns, than by the column ids in fColumns
+      bool operator< (const RInFlightCluster &other) const;
+   };
+
+   RPageSource *fPageSource;
+   /// The number of clusters before the currently active cluster that should stay in the pool if present
+   unsigned int fWindowPre;
+   /// The number of desired clusters in the pool, including the currently active cluster
+   unsigned int fWindowPost;
+   /// The cache of clusters around the currently active cluster
+   std::vector<std::shared_ptr<RCluster>> fPool;
+
+   std::mutex fLockInFlightClusters;
+   /// The clusters that were handed off to the I/O thread
+   std::vector<RInFlightCluster> fInFlightClusters;
+
+   std::mutex fLockWorkQueue;
+   std::condition_variable fCvHasWork;
+   /// The communication channel to the I/O thread
+   std::queue<RWorkItem> fWorkQueue;
+
+   std::thread fThreadIo;
+
+   /// Every cluster id has at most one corresponding RCluster pointer in the pool
+   std::shared_ptr<RCluster> FindInPool(DescriptorId_t clusterId);
+   /// Returns an index of an unused element in fPool; caller has to ensure that a free slot exists
```

Comment:
Added a bit more text to the docs.  The function asserts that there is a free spot (the application crashes otherwise).  The two methods calling it make sure there is a free spot by construction of the algorithm.